hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/fs/contract/AbstractContractCopyFromLocalTest.java

```diff
@@ -0,0 +1,285 @@
+        String updated = "updated";
+        FileUtils.write(file, updated, ASCII);
+        copyFromLocal(file, true);
+        assertFileTextEquals(dest, updated);
+    }
+
+    @Test
+    public void testCopyMissingFile() throws Throwable {
+        file = createTempFile("test");
+        file.delete();
+        // first upload to create
+        intercept(FileNotFoundException.class, "",
+                () -> copyFromLocal(file, true));
+    }
+
+    @Test
+    public void testSourceIsFileAndDelSrcTrue() throws Throwable {
+        describe("Source is a file delSrc flag is set to true");
+
+        file = createTempFile("test");
+        copyFromLocal(file, false, true);
+
+        assertFalse("uploaded file", Files.exists(file.toPath()));
+    }
+
+    @Test
+    public void testSourceIsFileAndDestinationIsDirectory() throws Throwable {
+        describe("Source is a file and destination is a directory. File" +
+                "should be copied inside the directory.");
+
+        file = createTempFile("test");
+        Path source = new Path(file.toURI());
+        FileSystem fs = getFileSystem();
+
+        File dir = createTempDirectory("test");
+        Path destination = fileToPath(dir);
+        fs.delete(destination, false);
+        mkdirs(destination);
+
+        fs.copyFromLocalFile(source, destination);
+        System.out.println("Did this work?");
+    }
+
+    @Test
+    public void testSrcIsDirWithFilesAndCopySuccessful() throws Throwable {
+        describe("Source is a directory with files, copy should copy all" +
+                " dir contents to source");
+        String firstChild = "childOne";
+        String secondChild = "childTwo";
+        File parent = createTempDirectory("parent");
+        File root = parent.getParentFile();
+        File childFile = createTempFile(parent, firstChild, firstChild);
+        File secondChildFile = createTempFile(parent, secondChild, secondChild);
+
+        copyFromLocal(parent, false);
+
+        assertPathExists("Parent directory not copied", fileToPath(parent));
+        assertFileTextEquals(fileToPath(childFile, root), firstChild);
+        assertFileTextEquals(fileToPath(secondChildFile, root), secondChild);
+    }
+
+    @Test
+    public void testSrcIsEmptyDirWithCopySuccessful() throws Throwable {
+        describe("Source is an empty directory, copy should succeed");
+        File source = createTempDirectory("source");
+        Path dest = copyFromLocal(source, false);
+
+        assertPathExists("Empty directory not copied", dest);
+    }
+
+    @Test
+    public void testSrcIsDirWithOverwriteOptions() throws Throwable {
+        describe("Source is a directory, destination exists and" +
+                "should be overwritten.");
+        File source = createTempDirectory("source");
+        String contents = "child file";
+        File child = createTempFile(source, "child", contents);
+
+        copyFromLocal(source, false);
+        // TODO: Fix local FS
```

Comment:
Not looking at a fix but clarification on how `copyFromLocal` should work with copying a source directory to a destination directory that's already there.
- LocalFS: from `dir/test.txt` to `dir/test.txt` => `dir/test.txt` and `dir/dir/test.txt`
- S3a: throws `PathExistsException`

Reply:
Thinking about it more, the test should be changed to:
- copy the source directory to destination;
- copy the source directory to the destination's parent which throws `PathExistsException` because of the first copy;
- retry the copy with overwrite set to `true` and check directory contents;

Reply:
Copy what HDFS does *unless it is clearly wrong and we are all confident that the change won't break things*. 
Given the experience of  HADOOP-13327 Output Stream Specification/#2587 I'm not confident about when we can do that any more.

If there's an overwrite flag, yes, it should allow overwrites.